### PR TITLE
Issue #4434 Implement context default request/response encodings.

### LINKED ISSUE #4434: Implement ServletContext.get/setRequestCharacterEncoding and ServletContext.get/setResponseCharacterEncoding
**Jetty 10**
Servlet 4.0

### PR DESCRIPTION
**Pull request by janbartel**
Implement the new servlet spec web.xml settings of request-character-encoding/response-character-encoding and equivalent ServletContext methods.

Closes #4434

### CODE REVIEW

File: jetty-server/src/main/java/org/eclipse/jetty/server/Response.java [hunk 1]
```diff
@@ -103,7 +104,7 @@
 
     private enum EncodingFrom
     {
-        NOT_SET, INFERRED, SET_LOCALE, SET_CONTENT_TYPE, SET_CHARACTER_ENCODING
+        NOT_SET, INFERRED, SET_LOCALE, SET_CONTENT_TYPE, SET_CHARACTER_ENCODING, DEFAULT
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/Response.java]: Should not the new constant be called `CONTEXT`?

File: jetty-server/src/test/java/org/eclipse/jetty/server/CharEncodingContextHandler.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.server;
+
```
Comment (sbordet) [inline jetty-server/src/test/java/org/eclipse/jetty/server/CharEncodingContextHandler.java:20]: Remove unnecessary newline.

File: jetty-server/src/test/java/org/eclipse/jetty/server/CharEncodingContextHandler.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+import org.eclipse.jetty.server.handler.ContextHandler;
+
+/**
+ * ContextHandler for testing that implements some
+ * CharEncoding methods from the servlet spec.
+ *
```
Comment (sbordet) [inline jetty-server/src/test/java/org/eclipse/jetty/server/CharEncodingContextHandler.java]: Remove empty javadoc line.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java [hunk 4]
```diff
@@ -177,6 +177,8 @@ public static void setServerInfo(String serverInfo)
     private final Map<String, String> _initParams;
     private ClassLoader _classLoader;
     private boolean _contextPathDefault = true;
+    private String _defaultRequestCharacterEncoding;
+    private String _defaultResponseCharacterEncoding;
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:181]: Is there a reason these fields and methods have "default" in their name?
I would just called them without "default" as they are in `ServletContext`.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-server/src/main/java/org/eclipse/jetty/server/Response.java (hunk 1 above)
Comment (janbartel) [reply jetty-server/src/main/java/org/eclipse/jetty/server/Response.java]: I'm removing the constant, as we never set the encoding on the response explicitly if it is via the context default.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java (hunk 4 above)
Comment (janbartel) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:181]: I'm disinclined to change the name. The "default" arises from the fact that it is the default for the context. Without "default" in the name, where the getters are accessed, it gets very confusing with so many different sources of character encoding flying around, so I like that it is obvious that it is the default only.

Comment (sbordet) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:181]: It is not the default for the context, it is the default for the requests or the responses. This is the only place "default" is used.
It is not used in `ServletContext`, it is not used in `web.xml`, it is not used in `StandardDescriptorProcessor`, and not even used in test method names, so I don't see any reason to change a name used in so many other places and add "default" (which really adds no additional clarity) in just this place.

**Comment by sbordet** [comment]
@janbartel besides the "default" discussion, the build failed?

**Comment by sbordet** [comment]
@janbartel I still don't like "default" and I would like it to be removed.

Other than that, do we need changes in `webdefault.xml`, explicitly adding these 2 elements?

**Comment by janbartel** [comment]
I don't understand your problem with having "default" in front of the name of the default request and response character encodings? That is what they are - defaults for the context. The servlet spec api is inconsistent: sometimes it uses "default" in method names eg getDefaultSessionTrackingModes() and web.xml elements eg default-context-path, sometimes it doesn't. However, the javadoc always makes clear that these are defaults and can be overridden by other means. I don't see why if the spec is inconsistent in naming we should be too in our own impl methods.

As to your question about webdefault.xml, I don't see why we would need to add to it?

**Comment by sbordet** [comment]
@janbartel it's not a problem, just I don't like it because it is only ever used in that one place for the encoding, all other places don't have "default" in it.

> As to your question about webdefault.xml, I don't see why we would need to add to it?

I was asking whether we must provide a default or not - apparently not.
`webdefault.xml` was the only other place I thought could be affected by this change, so I asked.

**Review by sbordet — APPROVED** [review]

**Pull request merged**